lib/puppet/transaction/event.rb

```diff
@@ -94,6 +98,30 @@ def to_yaml_properties
     YAML_ATTRIBUTES & super
   end
 
+  # Calculate and set the corrective_change parameter, based on the old_system_value of the property.
+  # @param [Object] old_system_value system_value from last transaction
+  # @return [bool] true if this is a corrective_change
+  def calculate_corrective_change(old_system_value)
+    # Only idempotent properties, and cases where we have an old system_value
+    # are corrective_changes.
+    if @property_instance.idempotent? &&
+       !@property_instance.sensitive &&
+       !old_system_value.nil?
+
+      # Here we duplicate the property to avoid side-effects, and modify its should value
+      # so we can use insync?
+      test_property = @property_instance.dup
```

Comment:
So its a squirrely one. I've incidentally now pushed this logic into the property so we can see it on its own. Let me point at that code to discuss this further.